File: app/src/main/java/xyz/hetula/homefy/setup/SetupFragment.java
/*
 * MIT License
 *
 * Copyright (c) 2017 Tuomo Heino
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

package xyz.hetula.homefy.setup;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import xyz.hetula.homefy.R;
import xyz.hetula.homefy.service.Homefy;
import xyz.hetula.homefy.service.protocol.VersionInfo;

public class SetupFragment extends Fragment {
    private static final String ADDRESS_KEY = "SetupFragement.ADDRESS_KEY";
    private static final String USERNAME_KEY = "SetupFragement.USERNAME_KEY";
    private static final String PASSWORD_KEY = "SetupFragement.PASSWORD_KEY";

    private View mViewCredentials;
    private Button mConnect;
    private EditText mAddress;
    private EditText mUser;
    private EditText mPass;
    private boolean mConnecting;
    private boolean mNeedsAuth = false;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        LinearLayout main = (LinearLayout) inflater.inflate(R.layout.fragment_setup, container, false);
        mConnect = (Button) main.findViewById(R.id.btn_connect);
        mConnect.setOnClickListener(this::onConClick);
        mAddress = (EditText) main.findViewById(R.id.txt_address);
        mUser = (EditText) main.findViewById(R.id.txt_username);
        mPass = (EditText) main.findViewById(R.id.txt_password);
        mViewCredentials = main.findViewById(R.id.view_credentials);
        mAddress.setOnEditorActionListener((v, actionId, event) -> {
            if (event != null && event.getKeyCode() == KeyEvent.KEYCODE_ENTER) {
                onConnect();
            }
            return false;
        });
        SharedPreferences pref = getActivity().getPreferences(Context.MODE_PRIVATE);
        mAddress.setText(pref.getString(ADDRESS_KEY, ""));
        mUser.setText(pref.getString(USERNAME_KEY, ""));
        mPass.setText(pref.getString(PASSWORD_KEY, ""));

        mConnecting = false;
        return main;
    }

    private void onConClick(View v) {
        if (mConnecting) return;
        String address = mAddress.getText().toString();
        Homefy.protocol().setServer(address);
        // Save it for later
        SharedPreferences pref = getActivity().getPreferences(Context.MODE_PRIVATE);
        pref.edit().putString(ADDRESS_KEY, address).apply();

        if (address.isEmpty()) {
            Toast.makeText(getContext(), "Enter Server address!", Toast.LENGTH_LONG).show();
            return;
        }
        if (!mNeedsAuth) {
            onConnect();
        } else {
            String user = mUser.getText().toString();
            String pass = mPass.getText().toString();
            if (user.isEmpty() || pass.isEmpty()) {
                Toast.makeText(getContext(), "Check Credentials!", Toast.LENGTH_LONG).show();
                return;
            }
            pref.edit()
                    .putString(USERNAME_KEY, user)
                    .putString(PASSWORD_KEY, pass)
                    .apply();
            onAuth(user, pass);
        }
    }

    private void onConnect() {
        mConnecting = true;
        Homefy.protocol().requestVersionInfo(this::onVersionInfo,
                volleyError -> {
                    Toast.makeText(getContext(),
                            "Error when Connecting!", Toast.LENGTH_LONG).show();
                    mConnecting = false;
                });
    }

    private void onAuth(String user, String pass) {
        mConnecting = true;
        Homefy.protocol().setAuth(user, pass);
        Homefy.protocol().requestVersionInfoAuth(this::onVersionInfoAuth,
                volleyError -> {
                    Toast.makeText(getContext(),
                            "Error when Connecting!", Toast.LENGTH_LONG).show();
                    mConnecting = false;
                });
    }

    private void onVersionInfo(VersionInfo versionInfo) {
        Log.d("SetupFragment", versionInfo.toString());
        mConnecting = false;
        switch (versionInfo.getAuthentication()) {
            case NONE:
                startLoading();
                break;
            case BASIC:
                mNeedsAuth = true;
                mConnect.setText(R.string.authenticate);
                mViewCredentials.setVisibility(View.VISIBLE);
                Toast.makeText(getContext(), "Enter Credentials!", Toast.LENGTH_LONG).show();
                break;
        }
    }

    private void onVersionInfoAuth(VersionInfo versionInfo) {
        Log.d("SetupFragment", versionInfo.toString());
        startLoading();
    }

    private void startLoading() {
        getFragmentManager()
                .beginTransaction()
                .replace(R.id.container, new LoadingFragment())
                .commit();
    }
}

File: app/src/main/kotlin/xyz.hetula.homefy/setup/SetupFragment.kt
/*
 * Copyright (c) 2018 Tuomo Heino
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package xyz.hetula.homefy.setup

import android.content.Context
import android.os.Bundle
import android.support.annotation.StringRes
import android.support.design.widget.Snackbar
import android.support.design.widget.TextInputEditText
import android.util.Log
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import xyz.hetula.homefy.HomefyFragment
import xyz.hetula.homefy.R
import xyz.hetula.homefy.service.protocol.RequestError
import xyz.hetula.homefy.service.protocol.VersionInfo


class SetupFragment : HomefyFragment() {
    private lateinit var mMain: LinearLayout
    private lateinit var mViewCredentials: View
    private lateinit var mConnect: Button
    private lateinit var mAddress: TextInputEditText
    private lateinit var mUser: TextInputEditText
    private lateinit var mPass: TextInputEditText
    private var mConnecting = false
    private var mNeedsAuth = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val main = inflater.inflate(R.layout.fragment_setup, container, false) as LinearLayout
        mMain = main
        mConnect = main.findViewById(R.id.btn_connect)
        mAddress = main.findViewById(R.id.txt_address)
        mUser = main.findViewById(R.id.txt_username)
        mPass = main.findViewById(R.id.txt_password)
        mViewCredentials = main.findViewById(R.id.view_credentials)

        mConnect.setOnClickListener { _ -> onConnectClick() }
        mAddress.setOnEditorActionListener { _, _, event ->
            if (event != null && event.keyCode == KeyEvent.KEYCODE_ENTER) {
                onConnect()
            }
            false
        }
        val pref = activity!!.getPreferences(Context.MODE_PRIVATE)
        mAddress.setText(pref.getString(ADDRESS_KEY, ""))
        mUser.setText(pref.getString(USERNAME_KEY, ""))
        mPass.setText(pref.getString(PASSWORD_KEY, ""))

        mConnecting = false
        return main
    }

    private fun onConnectClick() {
        if (mConnecting) return
        val address = mAddress.text.toString()
        homefy().getProtocol().server = address

        val pref = activity!!.getPreferences(Context.MODE_PRIVATE)
        pref.edit().putString(ADDRESS_KEY, address).apply()

        if (address.isEmpty()) {
            Snackbar.make(mMain, R.string.setup_enter_server_address, Snackbar.LENGTH_SHORT).show()
            return
        }
        if (!mNeedsAuth) {
            onConnect()
        } else {
            val user = mUser.text.toString()
            val pass = mPass.text.toString()
            if (user.isEmpty()) {
                Snackbar.make(mMain, R.string.setup_check_credentials_username, Snackbar.LENGTH_SHORT).show()
                return
            }
            if(pass.isEmpty()) {
                Snackbar.make(mMain, R.string.setup_check_credentials_password, Snackbar.LENGTH_SHORT).show()
                return
            }
            pref.edit()
                    .putString(USERNAME_KEY, user)
                    .putString(PASSWORD_KEY, pass)
                    .apply()
            onAuth(user, pass)
        }
    }

    private fun onConnect() {
        mConnecting = true
        homefy().getProtocol().requestVersionInfo(this::onVersionInfo,
                { err ->
                    Snackbar.make(mMain, getRequestErrorStringRes(err), Snackbar.LENGTH_SHORT).show()
                    mConnecting = false
                })
    }

    private fun onAuth(user: String, pass: String) {
        mConnecting = true
        homefy().getProtocol().setAuth(user, pass)
        homefy().getProtocol().requestVersionInfoAuth(this::onVersionInfoAuth,
                { err ->
                    Snackbar.make(mMain, getRequestErrorStringRes(err), Snackbar.LENGTH_SHORT).show()
                    mConnecting = false
                })
    }

    private fun onVersionInfo(versionInfo: VersionInfo) {
        Log.d("SetupFragment", versionInfo.toString())
        mConnecting = false
        when (versionInfo.authentication) {
            VersionInfo.AuthType.NONE -> startLoading()
            VersionInfo.AuthType.BASIC -> {
                mNeedsAuth = true
                mConnect.setText(R.string.setup_authenticate)
                mViewCredentials.visibility = View.VISIBLE
            }
            else -> {
                Log.w(TAG, "Unsupported auth method")
                Snackbar.make(mMain, R.string.setup_unsupported_authentication, Snackbar.LENGTH_LONG).show()
            }
        }
    }

    private fun onVersionInfoAuth(versionInfo: VersionInfo) {
        Log.d("SetupFragment", versionInfo.toString())
        mConnecting = false
        startLoading()
    }

    private fun startLoading() {
        fragmentManager!!
                .beginTransaction()
                .replace(R.id.container, LoadingFragment())
                .commit()
    }

    @StringRes
    private fun getRequestErrorStringRes(error: RequestError): Int = when(error.errCode) {
        401 -> R.string.setup_authentication_error
        -1 -> R.string.setup_malformed_url
        else -> R.string.setup_connection_error
    }

    companion object {
        private const val TAG = "SetupFragment"
        private const val ADDRESS_KEY = "SetupFragement.ADDRESS_KEY"
        private const val USERNAME_KEY = "SetupFragement.USERNAME_KEY"
        private const val PASSWORD_KEY = "SetupFragement.PASSWORD_KEY"
    }
}
